From 90af5fc57f409068ab5c0bf1dab313eb89736054 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Fri, 22 Dec 2023 18:09:20 +0800
Subject: [PATCH] hostapd: mtk: Add txpower vendor command

Add txpower vendor command including lpi, sku index and duplicate mode
control.
---
 hostapd/config_file.c             | 10 ++++++
 src/ap/ap_config.c                |  3 ++
 src/ap/ap_config.h                |  3 ++
 src/ap/ap_drv_ops.c               |  8 +++++
 src/ap/ap_drv_ops.h               |  1 +
 src/ap/hostapd.c                  |  2 ++
 src/ap/ieee802_11_he.c            |  3 ++
 src/common/mtk_vendor.h           | 14 +++++++++
 src/drivers/driver.h              |  9 ++++++
 src/drivers/driver_nl80211.c      | 51 +++++++++++++++++++++++++++++++
 src/drivers/driver_nl80211.h      |  1 +
 src/drivers/driver_nl80211_capa.c |  3 ++
 12 files changed, 108 insertions(+)

--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5112,6 +5112,16 @@ static int hostapd_config_fill(struct ho
 		u8 en = strtol(pos, NULL, 10);
 
 		conf->dfs_detect_mode = en;
+	} else if (os_strcmp(buf, "lpi_enable") == 0) {
+		u8 en = strtol(pos, NULL, 10);
+
+		conf->lpi_enable = !!en;
+	} else if (os_strcmp(buf, "sku_idx") == 0) {
+		conf->sku_idx = strtol(pos, NULL, 10);
+	} else if (os_strcmp(buf, "beacon_dup") == 0) {
+		u8 en = strtol(pos, NULL, 10);
+
+		conf->beacon_dup = !!en;
 	} else {
 		wpa_printf(MSG_ERROR,
 			   "Line %d: unknown configuration item '%s'",
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -307,6 +307,9 @@ struct hostapd_config * hostapd_config_d
 	conf->edcca_compensation = EDCCA_DEFAULT_COMPENSATION;
 	conf->three_wire_enable = THREE_WIRE_MODE_DISABLE;
 	conf->ibf_enable = IBF_DEFAULT_ENABLE;
+	conf->lpi_enable = 1;
+	conf->sku_idx = 0;
+	conf->beacon_dup = 1;
 
 	return conf;
 }
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1235,6 +1235,9 @@ struct hostapd_config {
 	u8 three_wire_enable;
 	u8 ibf_enable;
 	u8 dfs_detect_mode;
+	u8 lpi_enable;
+	u8 sku_idx;
+	u8 beacon_dup;
 };
 
 enum three_wire_mode {
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -1276,4 +1276,12 @@ int hostapd_drv_ibf_dump(struct hostapd_
 	if (!hapd->driver || !hapd->driver->ibf_dump)
 		return 0;
 	return hapd->driver->ibf_dump(hapd->drv_priv, ibf_enable);
-}
\ No newline at end of file
+}
+
+int hostapd_drv_txpower_ctrl(struct hostapd_data *hapd)
+{
+	if (!hapd->driver || !hapd->driver->txpower_ctrl)
+		return 0;
+	return hapd->driver->txpower_ctrl(hapd->drv_priv, hapd->iconf->lpi_enable,
+					  hapd->iconf->sku_idx, hapd->iconf->beacon_dup);
+}
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -153,6 +153,7 @@ int hostapd_drv_mu_dump(struct hostapd_d
 int hostapd_drv_three_wire_ctrl(struct hostapd_data *hapd);
 int hostapd_drv_ibf_ctrl(struct hostapd_data *hapd);
 int hostapd_drv_ibf_dump(struct hostapd_data *hapd, u8 *ibf_enable);
+int hostapd_drv_txpower_ctrl(struct hostapd_data *hapd);
 
 #include "drivers/driver.h"
 
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -2615,6 +2615,8 @@ dfs_offload:
 		goto fail;
 	if (hostapd_drv_ibf_ctrl(hapd) < 0)
 		goto fail;
+	if (hostapd_drv_txpower_ctrl(hapd) < 0)
+		goto fail;
 
 	wpa_printf(MSG_DEBUG, "%s: Setup of interface done.",
 		   iface->bss[0]->conf->iface);
--- a/src/ap/ieee802_11_he.c
+++ b/src/ap/ieee802_11_he.c
@@ -256,6 +256,8 @@ u8 * hostapd_eid_he_operation(struct hos
 			control = 3;
 		else
 			control = center_idx_to_bw_6ghz(seg0);
+		if (hapd->iconf->beacon_dup)
+			control |= HE_6GHZ_OPER_INFO_CTRL_DUP_BEACON;
 
 		control |= hapd->iconf->he_6ghz_reg_pwr_type <<
 			HE_6GHZ_OPER_INFO_CTRL_REG_INFO_SHIFT;
--- a/src/common/mtk_vendor.h
+++ b/src/common/mtk_vendor.h
@@ -15,6 +15,7 @@ enum mtk_nl80211_vendor_subcmds {
 	MTK_NL80211_VENDOR_SUBCMD_EDCCA_CTRL = 0xc7,
 	MTK_NL80211_VENDOR_SUBCMD_3WIRE_CTRL = 0xc8,
 	MTK_NL80211_VENDOR_SUBCMD_IBF_CTRL = 0xc9,
+	MTK_NL80211_VENDOR_SUBCMD_TXPOWER_CTRL = 0xce,
 };
 
 enum mtk_vendor_attr_edcca_ctrl {
@@ -241,6 +242,19 @@ ibf_dump_policy[NUM_MTK_VENDOR_ATTRS_IBF
 };
 
 
+enum mtk_vendor_attr_txpower_ctrl {
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_UNSPEC,
+
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_ENABLE,
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX,
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_BCN_DUP,
+
+	/* keep last */
+	NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL,
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_MAX =
+		NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL - 1
+};
+
 #define CSI_MAX_COUNT 256
 #define ETH_ALEN 6
 
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -5216,6 +5216,15 @@ struct wpa_driver_ops {
 	 *
 	 */
 	int (*ibf_dump)(void *priv, u8 *ibf_enable);
+
+	/**
+	* txpower_ctrl - ctrl txpower operation
+	* @priv: Private driver interface data
+	* @lpi_enable: 1 to enable lpi, 0 to disable lpi
+	* @beacon_dup: 1 to enable beacon duplicate, 0 to disable duplicate
+	* @sku_idx: index used to indicate which sku table should be used
+	*/
+	int (*txpower_ctrl)(void *priv, u8 lpi_enable, u8 sku_idx, u8 beacon_dup);
 };
 
 /**
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -88,6 +88,13 @@ static void handle_nl_debug_hook(struct
 	wpa_netlink_hook(tx, nlh, nlh->nlmsg_len);
 }
 
+static struct nla_policy
+txpower_ctrl_policy[NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL] = {
+	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_ENABLE] = { .type = NLA_U8 },
+	[MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX] = { .type = NLA_U8 },
+	[MTK_VENDOR_ATTR_TXPOWER_CTRL_BCN_DUP] = { .type = NLA_U8 },
+};
+
 static struct nl_sock * nl_create_handle(struct nl_cb *cb, const char *dbg)
 {
 	struct nl_sock *handle;
@@ -14375,6 +14382,49 @@ fail:
 	return -ENOBUFS;
 }
 
+static int nl80211_txpower_ctrl(void *priv, u8 lpi_enable, u8 sku_idx, u8 beacon_dup)
+{
+	struct i802_bss *bss = priv;
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	struct nl_msg *msg;
+	struct nlattr *data;
+	int ret;
+
+	if (!drv->mtk_txpower_vendor_cmd_avail) {
+		wpa_printf(MSG_INFO,
+			   "nl80211: Driver does not support setting lpi control");
+		return 0;
+	}
+
+	msg = nl80211_drv_msg(drv, 0, NL80211_CMD_VENDOR);
+	if (!msg)
+		goto fail;
+
+	if (nla_put_u32(msg, NL80211_ATTR_VENDOR_ID, OUI_MTK) ||
+		nla_put_u32(msg, NL80211_ATTR_VENDOR_SUBCMD,
+			    MTK_NL80211_VENDOR_SUBCMD_TXPOWER_CTRL))
+		goto fail;
+
+	data = nla_nest_start(msg, NL80211_ATTR_VENDOR_DATA);
+	if (!data)
+		goto fail;
+
+	nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_ENABLE, lpi_enable);
+	nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX, sku_idx);
+	nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_BCN_DUP, beacon_dup);
+
+	nla_nest_end(msg, data);
+	ret = send_and_recv(bss->drv->global, bss->drv->global->nl, msg, NULL, NULL, NULL, NULL, NULL);
+	if (ret)
+		wpa_printf(MSG_ERROR, "Failed to set lpi_enable. ret=%d (%s)", ret, strerror(-ret));
+
+	return ret;
+
+fail:
+	nlmsg_free(msg);
+	return -ENOBUFS;
+}
+
 const struct wpa_driver_ops wpa_driver_nl80211_ops = {
 	.name = "nl80211",
 	.desc = "Linux nl80211/cfg80211",
@@ -14541,4 +14591,5 @@ const struct wpa_driver_ops wpa_driver_n
 	.three_wire_ctrl = nl80211_enable_three_wire,
 	.ibf_ctrl = nl80211_ibf_enable,
 	.ibf_dump = nl80211_ibf_dump,
+	.txpower_ctrl = nl80211_txpower_ctrl,
 };
--- a/src/drivers/driver_nl80211.h
+++ b/src/drivers/driver_nl80211.h
@@ -203,6 +203,7 @@ struct wpa_driver_nl80211_data {
 	unsigned int mtk_mu_vendor_cmd_avail:1;
 	unsigned int mtk_3wire_vendor_cmd_avail:1;
 	unsigned int mtk_ibf_vendor_cmd_avail:1;
+	unsigned int mtk_txpower_vendor_cmd_avail:1;
 
 	u32 ignore_next_local_disconnect;
 	u32 ignore_next_local_deauth;
--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -1149,6 +1149,9 @@ static int wiphy_info_handler(struct nl_
 				case MTK_NL80211_VENDOR_SUBCMD_IBF_CTRL:
 					drv->mtk_ibf_vendor_cmd_avail = 1;
 					break;
+				case MTK_NL80211_VENDOR_SUBCMD_TXPOWER_CTRL:
+					drv->mtk_txpower_vendor_cmd_avail = 1;
+					break;
 				}
 			}
 
